### PR TITLE
Added expected regex for when an optional word may be needed in front…

### PR DESCRIPTION
**Pull request by spatel1009**
… of alternative group

cucumber-expression-pattern-test

## Summary

Added a test for the expected regex pattern for when an optional word may be needed in front of alternative non-capture group.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

**No changes or new code has been made towards the app**

## Checklist:

- [x] The change has been ported to Java.
- [ ] The change has been ported to Ruby.
- [ ] The change has been ported to JavaScript.
- [ ] The change has been ported to Go.
- [ ] The change has been ported to .NET.
- [x] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: cucumber-expressions/java/src/test/java/io/cucumber/cucumberexpressions/CucumberExpressionPatternTest.java [hunk 1]
```diff
@@ -35,6 +35,14 @@ public void translates_alternation_with_non_alpha() {
                 "^I said (?:Alpha1|Beta1)$"
         );
     }
+    
+    @Test
+    public void translates_alternation_with_optional_words() {
+        assertPattern(
+            "the (test )chat/call/email interactions are visible,
```
Comment (luke-hill) [inline cucumber-expressions/java/src/test/java/io/cucumber/cucumberexpressions/CucumberExpressionPatternTest.java]: we need a `"` at the end of this line, so there's no compilation error.

**Comment by luke-hill** [comment]
The error does look legit. But what's interesting is where it's tripping up is where we half expected it to.

Can you move the space one place back so instead of `(capture )` you have `( capture)` this should ensure a space between the right `)` and the start of the 3 alternated words. This would be a good one to see the output of.

**Comment by spatel1009** [comment]
> The error does look legit. But what's interesting is where it's tripping up is where we half expected it to.
> 
> Can you move the space one place back so instead of `(capture )` you have `( capture)` this should ensure a space between the right `)` and the start of the 3 alternated words. This would be a good one to see the output of.

Change it to `the( test) chat/call/email interactions are visible`?

**Comment by mpkorstanje** [comment]
It's ok. I found the problem already.

Given any expression of the form `(a )c/d`.

We first rewrite the optionals to regex so we get `(?:a )?c/d`.

Then we rewrite the alternatives so we get `(?:a (?:(?:)?c)|(?:d))` which happens to be a valid regex.

As you can see after rewriting the optionals we rewrite the alternatives. And because the rewritten optionals ends with ` )?` those last two characters are then rewritten into the group for the alternatives.

**Comment by luke-hill** [comment]
This is really where colours in markdown would help. Took me 15mins to figure out that regex. But yeh @spatel1009 that is right what you said. That should mitigate (But the original post is still causing some friction I agree)

**Comment by spatel1009** [comment]
Doing it that way with the( test) call/chat/email will only read a Gherkin written as `Then the test call/chat/email....` but not any of those singular like `Then the test call....` etc.  Also "test" is no longer be optional as steps like `Then the chat....` will not be read either.

**Comment by luke-hill** [comment]
I'm not sure I understand

`Then the( test) call/chat/email` will match

`Then the call`
`Then the chat`
`Then the test email`

**Comment by spatel1009** [comment]
Yes I thought so too but I get "cannot find declaration to go to" when using that above.  It reads none of the examples you have listed and the only one it reads is:
`Then xyza the test call/chat/email interactions page successfully loads with an interaction` 
Written exactly like that with all 3 alternative text options written with the /.

**Comment by luke-hill** [comment]
This feels like we're finding more edge cases as we go.

Ok noted! I'll have a little dabble and liaise up with @mpkorstanje to see if we can add tests for this also.

**Comment by mpkorstanje** [comment]
Superseded by #771
